File: frontend/lib/schemas/graph.ts
import { z } from "zod";

export const GraphNodeSchema = z.object({
  id: z.string(),
  name: z.string(),
  description: z.string(),
  content: z.string(),
  status: z.string(),
  tags: z.array(z.string()),
  supersedes: z.string().nullable(),
  supersededBy: z.string().nullable(),
  updatedAt: z.string(),
  createdAt: z.string(),
  connections: z.number().int(),
});

export const GraphLinkSchema = z.object({
  source: z.string(),
  target: z.string(),
  type: z.enum(["related", "supersedes"]),
});

export const GraphResponseSchema = z.object({
  nodes: z.array(GraphNodeSchema),
  links: z.array(GraphLinkSchema),
});

export type GraphNode = z.infer<typeof GraphNodeSchema>;
export type GraphLink = z.infer<typeof GraphLinkSchema>;
export type GraphResponse = z.infer<typeof GraphResponseSchema>;

export const UpdateNodeRequestSchema = z.object({
  name: z.string().min(1).max(120).optional(),
  description: z.string().optional(),
  content: z.string().optional(),
  status: z.string().optional(),
  tags: z.array(z.string()).optional(),
});

export type UpdateNodeRequest = z.infer<typeof UpdateNodeRequestSchema>;

File: backend/src/models/schemas/graph.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field

from src.models.schemas._base import MongoModel


class GraphNode(MongoModel):
    id: str = Field(validation_alias="_id")
    name: str
    description: str = Field(default="")
    content: str = Field(default="")
    status: str = Field(default="")
    tags: list[str] = Field(default_factory=list)
    supersedes: str | None = None
    superseded_by: str | None = Field(default=None, alias="supersededBy")
    updated_at: str = Field(default="", alias="updatedAt")
    created_at: str = Field(default="", alias="createdAt")
    connections: int = Field(default=0)


class UpdateNodeRequest(BaseModel):
    name: str | None = None
    description: str | None = None
    content: str | None = None
    status: str | None = None
    tags: list[str] | None = None


class UpdateNodeResponse(MongoModel):
    id: str = Field(validation_alias="_id")
    name: str
    description: str = ""
    content: str = ""
    status: str = ""
    tags: list[str] = Field(default_factory=list)
    supersedes: str | None = None
    superseded_by: str | None = Field(default=None, alias="supersededBy")
    updated_at: str = Field(default="", alias="updatedAt")
    created_at: str = Field(default="", alias="createdAt")
    connections: int = Field(default=0)


class GraphLink(BaseModel):
    source: str
    target: str
    type: Literal["related", "supersedes"]


class GraphResponse(BaseModel):
    nodes: list[GraphNode]
    links: list[GraphLink]
